### DISCUSSION TITLE
parameter settings

### DISCUSSION THREAD

**Discussion by kiminh** [Qihoo360/tensornet#65] [General]
I have trained a model on industrial dataset about one hundred million samples and 40+ feature slots.

I‘m wondering the parameter setting with industry best practices. @gzm55  @thundercatli 

maybe batch_size=64 is a better trade-off with such learning rate setting.

1. batch_size
the auc gap is very large( the smaller batch_size, the larger train auc), is this phenomenon correct?
 
```
batch_size = 1024
dense_opt = tn.core.Adam(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8)
sparse_opt = tn.core.AdaGrad(learning_rate=0.01, initial_g2sum=0.1, initial_scale=0.1)
columns[slot] = tn.feature_column.category_column(key=slot). bucket_size=1024 (default setting)
run time: 30 min+
log info: 43379/43379 [==============================] - 1839s 42ms/step - loss: 0.9954 - acc: 0.8117 - mse: 0.1441 - mae: 0.2654 - mape: 133282200.0000 - auc: 0.6809 - CTR: 0.1772 - PCTR: 0.1783 - COPC: 0.9936
```

```
batch_size = 256
dense_opt = tn.core.Adam(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8)
sparse_opt = tn.core.AdaGrad(learning_rate=0.01, initial_g2sum=0.1, initial_scale=0.1)
columns[slot] = tn.feature_column.category_column(key=slot). bucket_size=1024 (default setting)
run time: 1 hour+

log info:  162905/162905 [==============================] - 4157s 26ms/step - loss: 1.0213 - acc: 0.8250 - mse: 0.1299 - mae: 0.2584 - mape: 129333840.0000 - auc: 0.7327 - CTR: 0.1758 - PCTR: 0.1760 - COPC: 0.9986
```

```
batch_size = 128
dense_opt = tn.core.Adam(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8)
sparse_opt = tn.core.AdaGrad(learning_rate=0.01, initial_g2sum=0.1, initial_scale=0.1)
columns[slot] = tn.feature_column.category_column(key=slot). bucket_size=1024 (default setting)

run time: 1.5 hour+
log info:  328936/328936 [==============================] - 5501s 17ms/step - loss: 1.2424 - acc: 0.8256 - mse: 0.1277 - mae: 0.2548 - mape: 127114512.0000 - auc: 0.7504 - CTR: 0.1771 - PCTR: 0.1765 - COPC: 1.0031
```

```
batch_size = 32
dense_opt = tn.core.Adam(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8)
sparse_opt = tn.core.AdaGrad(learning_rate=0.01, initial_g2sum=0.1, initial_scale=0.1)
columns[slot] = tn.feature_column.category_column(key=slot). bucket_size=1024 (default setting)

run time: 7 hour +
log info: 1307371/1307371 [==============================] - 17070s 13ms/step - loss: 0.4293 - acc: 0.8296 - mse: 0.1210 - mae: 0.2406 - mape: 120372424.0000 - auc: 0.7875 - CTR: 0.1755 - PCTR: 0.1757 - COPC: 0.9991
```

seems the author noticed it https://github.com/Qihoo360/tensornet/issues/22, but with no feedback

let me follow the Linear Scaling Rule, namely
- When increasing batch size, proportionally increase the learning rate 

I will update the exp result here.

3. mpi run 
for local train, how to set the mpirun number, half of cpu numbers or what

**Comment by thundercatli** [comment]
1. You can try increase bucket_size of tn.feature_column.category_column for large batch size, you may need 40 * 1024 or more to save all signs for embedding lookup

2. Actually we don't use tn to do local train. You can check system metrics to see whether it's a memory or CPU bottleneck. Or use tensorboard with profiling. And you can try https://github.com/Qihoo360/hbox for tn job submit.

**Reply by kiminh** [reply]
> 1. You can try increase bucket_size of tn.feature_column.category_column for large batch size, you may need 40 * 1024 or more to save all signs for embedding lookup
> 2. Actually we don't use tn to do local train. You can check system metrics to see whether it's a memory or CPU bottleneck. Or use tensorboard with profiling. And you can try https://github.com/Qihoo360/hbox for tn job submit.

the default bucket size is 1024, which is equal larger the batch_size (either 1024 or 32). and category_column is set for every slot. then what do you mean to set large bucket_size (40*1024 = number of slots * batch_size)?  @thundercatli @gzm55 

https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/tensornet/feature_column/category_column.py#L76

```
def tn_category_columns_builder(features):
    """Builds tensornet category feature columns."""
    columns = {}
    for slot in set(features):
        columns[slot] = tn.feature_column.category_column(key=slot)
    return columns
```

**Reply by kiminh** [reply]
> 1. You can try increase bucket_size of tn.feature_column.category_column for large batch size, you may need 40 * 1024 or more to save all signs for embedding lookup
> 2. Actually we don't use tn to do local train. You can check system metrics to see whether it's a memory or CPU bottleneck. Or use tensorboard with profiling. And you can try https://github.com/Qihoo360/hbox for tn job submit.

I mean local train should only affect the run time of model train pipeline, not the model efficiency.
